### DISCUSSION TITLE
Universal attributes (like CSS)

### DISCUSSION THREAD

**Discussion by jack-jjm** [typst/typst#2605] [Ideas]
One of the things that's bugged me the most in trying out typst is how different attributes of text are segregated to different functions. This is very unlike the CSS model, in which attributes are universal. Let me explain - say we want to have an element which has a red border and red center aligned text. I start out with just my text:

```
#lorem(50)
```

Now I need to make it red. I look it up and find out the `text()` function has a text color attribute:

```
#text(fill: red)[
    #lorem(50)
]
```

Okay, now the red border. Coming from CSS I might try something like just passing an extra `border: 1px red` argument or something to `text()`, but of course that doesn't work. I look up the `text()` function and find it has no border parameter. Eventually I find the `block` function has one:

```
#block(stroke: (paint: red, thickness: 1pt), inset: 1em,
    text(fill: red)[
        #lorem(50)
    ]
)
```

And now to center-align the contents. Oh, the `block()` function doesn't even have an `align` attribute? Alright then:

```
#block(stroke: (paint: red, thickness: 1pt), inset: 1em,
    align(center,
        text(fill: red)[
            #lorem(50)
        ]
    )
)
```

Each new attribute I want to set introduces a new layer of nesting. The indentation hell can be mitigated by either not indenting every time or by using `set()` calls, but still, this is very verbose. Maybe more importantly, *I have to remember way more information*. With CSS I just have to remember the names of the attributes, here I have to remember the attributes *and* the mapping of attributes onto the functions that own them. And this mapping feels deeply arbitrary, and so is hard to remember.

I would much rather just write:

```
#format(
    stroke: (paint: red, thickness: 1pt),
    inset: 1em,
    text-color: red,
    align: center
)[
    #lorem(50)
]
```

Which is effectively the CSS model. Translating the Typst model to HTML + CSS, it would be as if each HTML tag had its own unique set of CSS attributes. Instead, in CSS we have a universal "pool" of attributes, and really all an HTML tag is (for the most part) is just a bag of default settings of those attributes, as a short hand. Except for things like `<video>` that have special behavior, you could get away with just one universal tag called `<tag>` purely as a hook for your CSS.

**Comment by laurmaedje** [comment]
Here's how I would write your example:

```typst
#block(stroke: 1pt + red, inset: 1em, {
  set align(center)
  set text(fill: red)
  lorem(50)
})
```

It's of course subjective, but I think this is clearer than the CSS-like alternative you posted. In the CSS model, there are often quirks because of the interaction of certain properties. For example, the `display` property can have a large effect on how other properties behave. It's essentially a quadratic amount of possibly interactions. Scoping properties to elements, in my opinion, gives them more well-defined behaviour. I actually also think that it improves discoverability because you can easily look through all closely related (e.g. text) properties. 

Another thing that is important in my opinion, is that set rules separate configuration of properties from creation of elements with that properties. For example, with your proposed `#format(stroke: (paint: red, thickness: 1pt))` I both create a stroke and configure it. I can't configure globally how my strokes behave. In contrast, with set rules, I can write: `#set underline(stroke: (paint: red, thickness: 1pt))` without actually creating an underline. To be able to do that, I would need a new `underline-stroke` global.

Last, but not least, I think the scoped approach is much more composable. We want to add user-defined types/elements to Typst in the future ([read more here](https://laurmaedje.github.io/posts/types-and-context/)). Set rules will naturally work with these, whereas global attributes would be sort of limited to built-in ones from the start.

**Comment by jack-jjm** [comment]
Would you mind just explaining how that example works exactly? In particular, how come we can't also just do `set block(stroke: ...)`? Why do we need to wrap with `block` but not with the other two?

**Reply by laurmaedje** [reply]
You could write `set block(stroke: ..)`, but you'd still need to create a `block[..]` separately. The general idea is that `set` only configures properties of elements that already exist, it never creates any new ones. `lorem(50)` creates a string which turns into a `text` element, which is affected by `set text`. 

The `block` is required because text itself has no ability to add a stroke around itself. In the spirit of separation of concerns, that's a problem of a separate container, in this case `block`.

I'll have to admit to `set align` is somewhat special: It not only affects the alignment of explicit `align` calls, but also of implicitly created `par` (automatic grouping of text, boxes, ..) and `flow` (automatic grouping of pars, blocks, ...) elements, simply because its too inconvenient otherwise. Similarly with `show ..: set block(above: ..)`. There is some inconsistency there, but I intend to rework how block spacing works overall because it's not really intuitive.
